Refactor logical operator constraints

ptomulik opened the pull request:
I'd like to propose the following changes to the implementation of logical-operator constraints, namely ``LogicalOr``, ``LogicalAnd``, ``LogicalXor`` and ``LogicalNot``. My initial intent was to implement a functionality that would allow externally developed constraints to provide meaningful messages when they appear in ``LogicalNot`` expression, but it quickly appeared that this needs more attention than just a quick fix.

The PR introduces the following changes:

- new abstract base classes for logical operators, the hierarchy is now (**edited**):

  ```console
  Constraint          (abstract, extended)
  + Operator          (abstract, new)
     + BinaryOperator (abstract, new)
       + LogicalAnd   (final, changed)
       + LogicalOr    (final, changed)
       + LogicalXor   (final, changed)
     + UnaryOperator  (abstract, new)
       + LogicalNot   (final, changed)
  ```
- introduced concept of operator's ``arity`` and ``precedence``; the corresponding methods are

  ```php
    Operator::arity(): int;
    Operator::precedence(): int;
  ```
- if an ``Operator`` becomes an operand to another operator, its ``arity`` and ``precedence`` are taken into account when forming message strings, especially w.r.t. wrapping with parentheses, for example

  ```php
        $constraint = LogicalAnd::fromConstraints(
            LogicalOr::fromConstraints(new IsType('bool'), new IsType('int')),
            LogicalOr::fromConstraints(new IsIdentical(false), new IsIdentical(0))
        );
        echo $constraint->toString() . "\n";
  ```
  yields

  ```console
    ( is of type "bool" or is of type "int" ) and ( is identical to false or is identical to 0 )
  ```

  while

  ```php
        $constraint = LogicalOr::fromConstraints(
            LogicalAnd::fromConstraints(new IsType('bool'), new IsIdentical(false)),
            LogicalAnd::fromConstraints(new IsType('int'), new IsIdentical(0))
        );
        echo $constraint->toString() . "\n";
  ```

  yields

  ```console
  is of type "bool" and is identical to false or is of type "int" and is identical to 0
  ```
  This shall work with all properly implemented sub-classes of ``Operator``.

- when only a single constraint is provided as operand to a binary operator (``LogicalAnd``, ``LogicalOr``, ``LogicalXor``), the binary operator becomes "degenerate", for example

  ```php
        $constraint = LogicalOr::fromConstraints(new IsTrue());
        $constraint->evaluate(false);  // same effect as for (new IsTrue())->evaluate(false);
  ```

  yields

  ```console
  PHP Fatal error:  Uncaught Failed asserting that false is true.
  ```

- all three binary operators now share same implementation of ``setConstraints()`` method, which accepts non-constraint values and decorates them with ``IsEqual``; previously only ``LogicalOr`` and ``LogicalXor`` worked this way, while ``LogicalAnd`` was refusing non-constraint values by throwing an exception,

- when a binary operator is configured with no operands, it always evaluates to ``false`` (always fails); previously ``LogicalXor`` and ``LogicalAnd`` were returning ``true``, while ``LogicalOr`` was returning ``false``,

- two methods are added to ``Constraint`` class

  ```php
    Constraint::toStringInContext(Operator $operator, int $position): string;
    Constraint::failureDescriptionInContext(Operator $operator, int $position, $other): string;
  ```

  newly implemented constraints may override the default implementation of ``toStringInContext()`` to provide meaningful messages when in context of ``LogicalNot`` operator or such, for example:

  ```php
    class ImplementsInterface extends Constraint
    {
        protected $expected;

        public function __construct(string $expected)
        {
            $this->expected = $expected;
        }

        public function toString(): string
        {
            return 'implements interface';
        }

        public function toStringInContext(Operator $operator, int $position): string
        {
            switch (get_class($operator)) {
                case LogicalNot::class:
                    return 'does not implement interface';
                default:
                    return '';
            }
        }

        protected function matches($other): bool
        {
            return (new \ReflectionClass($other))->implementsInterface($this->expected);
        }
    }

    $constraint = new ImplementsInterface(\Throwable::class);
    echo (new LogicalNot($constraint))->toString()."\n";
  ```
  
  which yields

  ```console
  does not implement interface
  ```

- the original ``Constraint::evaluate()`` is now used; particular operators only override ``matches()`` method; this reduced code duplication, however the $description argument is no longer passed down to operands; this seems to be not an issue, at least in scope of PHPUnit, as operators were calling ``$constraint->evaluate($other, $description, true)`` and with the default implementation of ``Constraint::evaluate``, as well as with all its overrides I've noticed in PHPunit, this call is equivalent to ``$constraint->evaluate($other, '', true)``, which is how it's done now, 

- re-implemented unit tests; most of the actual testing code is now placed in new abstract base classes, because the actual test code for particular operators appeared to differ only in small details like operator name, arity, precedence, truth table; the hierarchy of testing classes is now

  ```console
    TestCase                     (abstract, unaltered)
   + OperatorTestCase            (abstract, new)
      + BinaryOperatorTestCase   (abstract, new)
        + LogicalAndTest         (final, changed)
        + LogicalOrTest          (final, changed)
        + LogicalXorTest         (final, changed)
      + UnaryOperatorTestCase    (abstract, new)
        + LogicalNotTest         (final, changed)
  ```

I hope this will be useful for the project. One of my projects is actually awaiting for this extension.

ptomulik committed:
Logical operators revised

* refactored class hierarchy to distinguish logical operators,
  among regular constraints, new abstract classes introduced,
  so the current hierarchy is:

        Constraint      (abstract)
        + Operator      (abstract)
          + Connective  (abstract)
            + LogicalAnd
            + LogicalOr
            + LogicalXor
          + Unary       (abstract)
            + LogicalNot

* rewrote unit tests for logical operators, most of the testing code is
  written into abstract *TestCase classes, the hierarchy is following:

        TestCase
        + ConstraintTestCase        (abstract)
          + OperatorTestCase        (abstract)
            + ConnectiveTestCase    (abstract)
              + LogicalAndTest
              + LogicalOrTest
              + LogicalXorTest
            + UnaryTestCase         (abstract)
              + LogicalNotTest

* add two methods to Constraint class to allow returning different
  messages in different contexts (e.g. when a constraint appears to be
  an argument to LogicalNot constraint), the methods are:

        toStringInContext()
        failureDescriptionInContext()

* logical operators no logner pass their messages to their operands,
  seems like the message was never used within the evaluate() method
  of operand constraints, at least across PHPUnit code, this change
  allowed to reduce code redundancy and simplified the implementation
  a little bit.

sebastianbergmann commented:
Just to be clear: this does not replace (because it includes) #4251?

sebastianbergmann commented:
Although `roave-backward-compatibility-check` did not detect any change that breaks backward compatibility (which is good!), I will need to carefully review this.

sebastianbergmann commented:
I understand what `Connective`, `Unary`, and `Operator` mean but I fear these names may be too succinct. What about

* `LogicalOperatorConstraint` instead of `Operator`
* `ConnectiveLogicalOperatorConstraint` instead of `Connective`
* UnaryLogicalOperatorConstraint` instead of `Unary`

sebastianbergmann commented:
By the way: thank you for working on this! This dark corner of PHPUnit's codebase has been in need of a proper cleanup for way too long.

ptomulik commented:
> Just to be clear: this does not replace (because it includes) #4251?

It includes fix for #4251.

ptomulik commented:
> I understand what `Connective`, `Unary`, and `Operator` mean but I fear these names may be too succinct. What about
> 
>     * `LogicalOperatorConstraint` instead of `Operator`
> 
>     * `ConnectiveLogicalOperatorConstraint` instead of `Connective`
> 
>     * `UnaryLogicalOperatorConstraint` instead of `Unary`
> 
> 
> ?

I was thinking about it, and stopped at these conclusions:

- these classes land in ``Constraint`` namespace, thus the ``*Constraint`` suffix seemed quite redundant; actually anything that is found under the ``Constraint`` namespace is IMHO silently assumed to be a ``Constraint``?,
- at first look, the only meaningful operators on operands being ``Contraint``s, are logical operators; in essence they operate on boolean values returned by operands and return boolean results, so ``Logical`` prefix seemed little bit unnecessary, but
- even if someone invents an operator that takes one or more ``Constraint``s but is not a classical logical operator (for example, an operator which compares number of failing operands against some integer), she should still be able to reuse these abstract classes to implement her own "not a logical" operator; there is nothing in these abstract classes that would restrict their reuse to logical operators, at lest I can't see such factors,

Specific features of these classes are:
- ``Operator``:
    - has operator name,
    - has precedence,
    - has arity,
- ``Unary``
    - its arity is always 1,
    - by default, argument is on the right side of the operator (this can be overriden in a subclass),
- ``Connective``
   - its arity is from 0 to inf,
   - by default uses infix notation (this can be overriden in a subclass),

That's why decided for short names, simply couldn't imagine other objects whose names could possibly collide, at least within this namespace.

ptomulik commented:
I'm going to revise the logic for automatic wrapping subexpressions with  parenheses. The current approach just revealed some problems. Will append new commits with tests and fixes.

ptomulik committed:
fixing parenthesing logic - unary operators #1

ptomulik committed:
added tests for LogicalNot::negate()

ptomulik committed:
fixed LogicalNot::negate() to prevent replacing subwords

ptomulik committed:
refactored ConnectiveTestCase

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+            8,
+        ];
+
+        $constraints = \array_map(function ($count) {
```
```suggestion
        $constraints = \array_map(static function ($count): \CountConstraint {
```

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+     */
+    public function testEvaluateReturnsCorrectBooleanResult(array $inputs, bool $expected): void
+    {
+        $constraints = \array_map(function (bool $input) {
```
```suggestion
        $constraints = \array_map(static function (bool $input): \BooleanConstraint {
```

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+
+        $constraint = $className::fromConstraints(...$constraints);
+
+        if ($expected) {
+            $this->assertNull($constraint->evaluate(null));
+        } else {
+            $expectedString = self::operatorJoinStrings(
+                \array_map(
+                    function (Constraint $operand) {
+                        return $operand->toString();
+                    },
+                    $constraints
+                )
+            );
+            $message = "Failed asserting that 'the following expression is true' " . $expectedString;
+            $this->expectException(ExpectationFailedException::class);
+            $this->expectExceptionMessage($message);
+            $constraint->evaluate('the following expression is true');
+        }
```
```suggestion
        if ($expected) {
            $this->assertNull($constraint->evaluate(null));

            return;
        }

        $expectedString = self::operatorJoinStrings(\array_map(static function (Constraint $operand): string {
            return $operand->toString();
        }, $constraints));

        $message = "Failed asserting that 'the following expression is true' " . $expectedString;
        $this->expectException(ExpectationFailedException::class);
        $this->expectExceptionMessage($message);
        $constraint->evaluate('the following expression is true');
```

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+     */
+    public function testToStringWithNamedConstraints(string ...$names): void
+    {
+        $constraints = \array_map(function (string $name) {
```
```suggestion
        $constraints = \array_map(static function (string $name): \NamedConstraint {
```

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+
+        $className = $this->className();
+
+        $operators = \array_map(function (Constraint $operand) use ($className) {
```
```suggestion
        $operators = \array_map(static function (Constraint $operand) use ($className): Constraint {
```

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+
+        $className = $this->className();
+
+        $operators = \array_map(function (Constraint $operand) use ($className) {
```
```suggestion
        $operators = \array_map(static function (Constraint $operand) use ($className): Constraint {
```

localheinz commented on tests/unit/Framework/Constraint/ConnectiveTestCase.php:
```diff
@@ -0,0 +1,604 @@
+    {
+        $tuples = self::getBinaryTuples($minSize, $maxSize);
+
+        return \array_map(function ($tuple) {
```
```suggestion
        return \array_map(static function ($tuple): array {
```

localheinz commented on src/Framework/Constraint/Connective.php:
```diff
@@ -0,0 +1,113 @@
+ *
+ * Infix notation of the operator is assumed.
+ */
+abstract class Connective extends Operator
```
Not so sure about the naming and the inheritance here, perhaps something along the lines of `CompositeConstraint` would be better?

What do you think?

ptomulik replied:
``CompositeConstraint`` does not reveal the actual purpose of these objects (it only emphasizes composition). The origin of the ``Connective`` class name is [Logical Connective](https://en.wikipedia.org/wiki/Logical_connective), which denotes a logical operator (binary or unary, Wikipedia says). I adopted this name for operators that accept arbitrary number of arguments (0 to inf). We can still look for a better name.

sebastianbergmann replied:
Then `LogicalConnective` is probably the name we're looking for. After all, we already have `LogicalAnd`, etc.

ptomulik replied:
With new commits, I've renamed base classes to the following: ``Unary`` -> ``UnaryOperator``, ``Connective`` -> ``BinaryOperator``. I'll set this as resolved for now, but we can get back to it and rename these classes further.

ptomulik commented on src/Framework/Constraint/Connective.php:
```diff
@@ -0,0 +1,113 @@
+     */
+    private $constraints = [];
+
+    public static function fromConstraints(Constraint ...$constraints): self
```
I think, we shall leave (or provide) at least one point of extension which receives the array of arguments provided for construction. Custom operators may need to evaluate these inputs in their own manner (for example, they may need to restrict number of arguments). While making this method final may be OK, the job of setting constraints could be delegated to ``setConstraints()``.

ptomulik commented on src/Framework/Constraint/Connective.php:
```diff
@@ -0,0 +1,113 @@
+    /**
+     * @param mixed[] $constraints
+     */
+    public function setConstraints(array $constraints): void
```
If this has to be final, then we should provide implementors of custom operators with additional method that accepts/validates the array of constraints as a whole (currently there is only a method for individual arguments).

ptomulik commented on src/Framework/Constraint/Constraint.php:
```diff
@@ -152,4 +152,52 @@ protected function failureDescription($other): string
+     * @param Operator $operator the $operator of the expression
+     * @param int      $position position in $operator expression
+     */
+    protected function toStringInContext(Operator $operator, int $position): ?string
```
I interpret null as "not implemented, please use toString() instead" in several places. Generally, empty string is a special case of the description string. Maybe there is another way to pass this information.

sebastianbergmann replied:
Would raising an exception make sense when the `Constraint` cannot be represented as a string?

ptomulik replied:
Well, then the default behavior will be to raise exception. Maybe my wording was quite misleading, by "not implemented" I meant that the sting is not specialized for the context of given $operator and should be taken from ``toString()`` instead. I'll see if it's possible to refactor it such that the function fallbacks to ``toString()`` by itself. If it's possible, then it will always return string.

ptomulik commented on src/Framework/Constraint/Constraint.php:
```diff
@@ -152,4 +152,52 @@ protected function failureDescription($other): string
+     * @param int      $position position in $operator expression
+     * @param mixed    $other    evaluated value or object
+     */
+    protected function failureDescriptionInContext(Operator $operator, int $position, $other): ?string
```
I interpret null as "not implemented, please use failureDescription() instead". Generally, empty string is a special case of the description string. Maybe there is another way to pass this information.

ptomulik commented on src/Framework/Constraint/LogicalAnd.php:
```diff
@@ -9,112 +9,42 @@
-use PHPUnit\Framework\ExpectationFailedException;
-
 /**
  * Logical AND.
```
Ok.

ptomulik commented on src/Framework/Constraint/LogicalNot.php:
```diff
@@ -9,18 +9,11 @@
-use PHPUnit\Framework\ExpectationFailedException;
-
 /**
  * Logical NOT.
```
Ok.

ptomulik commented on src/Framework/Constraint/LogicalOr.php:
```diff
@@ -9,108 +9,42 @@
-use PHPUnit\Framework\ExpectationFailedException;
-
 /**
  * Logical OR.
```
Ok.

ptomulik commented on src/Framework/Constraint/LogicalXor.php:
```diff
@@ -9,115 +9,49 @@
-use PHPUnit\Framework\ExpectationFailedException;
-
 /**
  * Logical XOR.
```
Ok.

ptomulik commented on src/Framework/Constraint/Operator.php:
```diff
@@ -0,0 +1,54 @@
+namespace PHPUnit\Framework\Constraint;
+
+/**
+ * Abstract base class for logical operators.
```
Maybe not. I'll clean it up.

ptomulik replied:
Resolved manually.

ptomulik commented on src/Framework/Constraint/Operator.php:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Validates $constraint argument.
+     */
+    protected function checkConstraint($constraint): Constraint
```
I'd rather leave it non-final. Custom constraints may need to override this (and, obviously, the corresponding unit test).

ptomulik commented on src/Framework/Constraint/Operator.php:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Validates $constraint argument.
+     */
+    protected function checkConstraint($constraint): Constraint
+    {
+        if (!($constraint instanceof Constraint)) {
+            $constraint = new IsEqual($constraint);
+        }
+
+        return $constraint;
+    }
```
I'd rather leave the method non-final. The rest of the patch is OK to me.

ptomulik replied:
I've fixed the function body manually as suggested. The signature was left unchanged.

ptomulik commented on src/Framework/Constraint/Unary.php:
```diff
@@ -0,0 +1,108 @@
+namespace PHPUnit\Framework\Constraint;
+
+/**
+ * Abstract base class for unary logical operators.
```
Not necesarilly. Will remove it.

ptomulik commented on src/Framework/Constraint/Unary.php:
```diff
@@ -0,0 +1,108 @@
+
+        $string = $this->constraint->toStringInContext($this, 1);
+
+        return $string ?? $this->transformString(
+            $this->constraint->toString()
+        );
```
Looks like line 47 is duplicated in line 49. Otherwise, it's OK to me.

ptomulik replied:
Resolved manually.

ptomulik commented on tests/_files/BooleanConstraint.php:
```diff
@@ -0,0 +1,37 @@
+
+    public function toString(): string
+    {
+        return $this->matches ? 'true' : 'false';
```
Isn't the expression too simple to split it into four lines of code?

ptomulik committed:
Apply suggestions from code review by localheinz

Co-authored-by: Andreas Möller <am@localheinz.com>

ptomulik committed:
Manual resolutions after code review

ptomulik committed:
Apply suggestions from code review - second pass

Co-authored-by: Andreas Möller <am@localheinz.com>

ptomulik committed:
Applied php-cs-fixer

ptomulik commented on tests/unit/Framework/Constraint/UnaryTestCase.php:
```diff
@@ -0,0 +1,626 @@
+        ));
+    }
+
+    public function testOperatorName(): void
```
Maybe most, not necessarily all.

ptomulik committed:
Apply suggestions from code review - pass 3

Co-authored-by: Andreas Möller <am@localheinz.com>

ptomulik committed:
Applied php-cs-fixer

ptomulik committed:
Implemented Constraint::reduce() and renamed operator abstract classes

sebastianbergmann commented:
Your latest changes are not backward compatible:
```
[BC] CHANGED: Method fromConstraints() of class PHPUnit\Framework\Constraint\LogicalOr became final
[BC] CHANGED: Method fromConstraints() of class PHPUnit\Framework\Constraint\LogicalOr became final
[BC] CHANGED: Method setConstraints() of class PHPUnit\Framework\Constraint\LogicalOr became final
[BC] CHANGED: Method toString() of class PHPUnit\Framework\Constraint\LogicalOr became final
[BC] CHANGED: Method count() of class PHPUnit\Framework\Constraint\LogicalOr became final
[BC] CHANGED: Method fromConstraints() of class PHPUnit\Framework\Constraint\LogicalAnd became final
[BC] CHANGED: Method fromConstraints() of class PHPUnit\Framework\Constraint\LogicalAnd became final
[BC] CHANGED: Method setConstraints() of class PHPUnit\Framework\Constraint\LogicalAnd became final
[BC] CHANGED: Method toString() of class PHPUnit\Framework\Constraint\LogicalAnd became final
[BC] CHANGED: Method count() of class PHPUnit\Framework\Constraint\LogicalAnd became final
[BC] CHANGED: Method fromConstraints() of class PHPUnit\Framework\Constraint\LogicalXor became final
[BC] CHANGED: Method fromConstraints() of class PHPUnit\Framework\Constraint\LogicalXor became final
[BC] CHANGED: Method setConstraints() of class PHPUnit\Framework\Constraint\LogicalXor became final
[BC] CHANGED: Method toString() of class PHPUnit\Framework\Constraint\LogicalXor became final
[BC] CHANGED: Method count() of class PHPUnit\Framework\Constraint\LogicalXor became final
15 backwards-incompatible changes detected
```

I agree that these classes should be `final`. We will have to wait until PHPUnit 10, though, to declare them `final`.

ptomulik committed:
Reverted backward-incompatible changes in Connective.php

ptomulik committed:
Fixed attributes for some *.php files

ptomulik committed:
Try to convince type checker

ptomulik committed:
Changed toStringInContext() and failureDescriptionInContext() to always return string

ptomulik commented:
I think, this is good checkpoint for next review. What's changed since the last review is:

- renamed ``Unary`` -> ``UnaryOperator``, and ``Connective`` -> ``BinaryOperator``,
- fixed ``LogicalNot::negate()`` to not replace half-words (previously it tended to replace ``this`` with ``this not``, ``coexists`` with ``co does not exists`` and so on),
- changed ``toStringInContext()`` and ``failureDescriptionInContext()`` to return empty string instead of ``null``,
- introduced a concept of "degenerate subexpressions" and expression reductions (new method ``Constraint::reduce()``), which appears quite important if we want to generate correct strings with ``toString()`` and ``failureDescription()``,
- added new unit test named ``LogicalExpressionsTest.php``, this is not related to any particular PHP class, but it's the place where we test combinations of logical operators (expressions made of PHPUnit Constraints and Operators).

In conclusion I'd like to share my observation, that the whole hierarchy of Constraints, Operators etc. could be organized in form of an abstract/concreate syntax tree (as it happends in parsers), to resemble a syntax involving expressions (including operator expressions) and terminal symbols (leaf nodes - curently non-operator Constraints). Some thoughts are shared in [this comment](https://github.com/ptomulik/phpunit/blob/028668dd87ec720b51623ea4340d783a1a2f3e74/src/Framework/Constraint/Constraint.php#L203).

ptomulik commented:
I created an experimental branch, based on ``master`` but the whole ``src`` directory is taken from this branch. So it has all tests from ``master``, but new source code. I've tested the new code with old unit tests, the log is included below. The mixed branch is available [here](https://github.com/ptomulik/phpunit/commits/logical-operators-revised-with-old-phpunit). The two failures below are due to:

1) changed behavior of ``LogicalAnd::setConstraints()`` - now it accepts non-constraint arguments and decorate them with ``IsEqual`` constraint, previously it was throwing an exception.
2) fixed ``LogicarXor`` implementation, which returns wrong result in ``master``. The assertion at ``/home/ptomulik/sources/phpunit/tests/unit/Framework/Constraint/LogicalXorTest.php:45`` is simply wrong in ``master``.

```console
PHPUnit 9.3-g257cf469d by Sebastian Bergmann and contributors.

Runtime:       PHP 7.3.15-3 with Xdebug 2.7.2
Configuration: /home/ptomulik/sources/phpunit/phpunit.xml

.............................................................   61 / 2254 (  2%)
.............................................................  122 / 2254 (  5%)
.............................................................  183 / 2254 (  8%)
.............................................................  244 / 2254 ( 10%)
.............................................................  305 / 2254 ( 13%)
.............................................................  366 / 2254 ( 16%)
.............................................................  427 / 2254 ( 18%)
.............................................................  488 / 2254 ( 21%)
.............................................................  549 / 2254 ( 24%)
.............................................................  610 / 2254 ( 27%)
.............................................................  671 / 2254 ( 29%)
.............................................................  732 / 2254 ( 32%)
.............................................................  793 / 2254 ( 35%)
.............................................F...............  854 / 2254 ( 37%)
...............F.............................................  915 / 2254 ( 40%)
.............................................................  976 / 2254 ( 43%)
............................................................. 1037 / 2254 ( 46%)
............................................................. 1098 / 2254 ( 48%)
...........................................................SS 1159 / 2254 ( 51%)
SS............SS............................................. 1220 / 2254 ( 54%)
............................................................. 1281 / 2254 ( 56%)
............................................................. 1342 / 2254 ( 59%)
............................................................. 1403 / 2254 ( 62%)
............................................................. 1464 / 2254 ( 64%)
............................................................. 1525 / 2254 ( 67%)
............................................................. 1586 / 2254 ( 70%)
............................................................. 1647 / 2254 ( 73%)
............................................................. 1708 / 2254 ( 75%)
............................................................. 1769 / 2254 ( 78%)
............................................................. 1830 / 2254 ( 81%)
............................................................. 1891 / 2254 ( 83%)
............................................................. 1952 / 2254 ( 86%)
.......................S..................................... 2013 / 2254 ( 89%)
............................................................. 2074 / 2254 ( 92%)
...............S............................................. 2135 / 2254 ( 94%)
.SSSSS....................................................S.. 2196 / 2254 ( 97%)
..........................................................    2254 / 2254 (100%)

Time: 00:36.985, Memory: 32.00 MB

There were 2 failures:

1) PHPUnit\Framework\Constraint\LogicalAndTest::testSetConstraintsRejectsInvalidConstraint
Failed asserting that exception of type "PHPUnit\Framework\Exception" is thrown.

/home/ptomulik/sources/phpunit/src/Framework/Constraint/Constraint.php:119
/home/ptomulik/sources/phpunit/src/Framework/Constraint/Constraint.php:53
/home/ptomulik/sources/phpunit/src/Framework/Assert.php:2228
/home/ptomulik/sources/phpunit/src/Framework/TestCase.php:1461
/home/ptomulik/sources/phpunit/src/Framework/TestCase.php:1029
/home/ptomulik/sources/phpunit/src/Framework/TestResult.php:691
/home/ptomulik/sources/phpunit/src/Framework/TestCase.php:770
/home/ptomulik/sources/phpunit/src/Framework/TestSuite.php:637
/home/ptomulik/sources/phpunit/src/Framework/TestSuite.php:637
/home/ptomulik/sources/phpunit/src/Framework/TestSuite.php:637
/home/ptomulik/sources/phpunit/src/TextUI/TestRunner.php:651
/home/ptomulik/sources/phpunit/src/TextUI/Command.php:108
/home/ptomulik/sources/phpunit/src/TextUI/Command.php:68

2) Framework\Constraint\LogicalXorTest::testFromConstraintsReturnsConstraint
Failed asserting that false is true.

/home/ptomulik/sources/phpunit/src/Framework/Constraint/Constraint.php:119
/home/ptomulik/sources/phpunit/src/Framework/Constraint/Constraint.php:53
/home/ptomulik/sources/phpunit/src/Framework/Assert.php:2228
/home/ptomulik/sources/phpunit/src/Framework/Assert.php:1022
/home/ptomulik/sources/phpunit/tests/unit/Framework/Constraint/LogicalXorTest.php:45
/home/ptomulik/sources/phpunit/src/Framework/TestCase.php:1412
/home/ptomulik/sources/phpunit/src/Framework/TestCase.php:1029
/home/ptomulik/sources/phpunit/src/Framework/TestResult.php:691
/home/ptomulik/sources/phpunit/src/Framework/TestCase.php:770
/home/ptomulik/sources/phpunit/src/Framework/TestSuite.php:637
/home/ptomulik/sources/phpunit/src/Framework/TestSuite.php:637
/home/ptomulik/sources/phpunit/src/Framework/TestSuite.php:637
/home/ptomulik/sources/phpunit/src/TextUI/TestRunner.php:651
/home/ptomulik/sources/phpunit/src/TextUI/Command.php:108
/home/ptomulik/sources/phpunit/src/TextUI/Command.php:68

--

There were 14 skipped tests:

[...]

FAILURES!
Tests: 2254, Assertions: 3809, Failures: 2, Skipped: 14.
```
